src/front/glsl/variables.rs

```diff
@@ -51,141 +122,125 @@ impl Parser {
-                    width: 4,
+                builtin: match name {
+                    "gl_GlobalInvocationID" => BuiltIn::GlobalInvocationId,
+                    "gl_NumWorkGroups" => BuiltIn::NumWorkGroups,
+                    "gl_WorkGroupSize" => BuiltIn::WorkGroupSize,
+                    "gl_WorkGroupID" => BuiltIn::WorkGroupId,
+                    "gl_LocalInvocationID" => BuiltIn::LocalInvocationId,
+                    _ => unreachable!(),
                 },
-                BuiltIn::NumWorkGroups,
-                false,
-                StorageQualifier::Input,
-            ),
-            "gl_FrontFacing" => add_builtin(
-                TypeInner::Scalar {
+                mutable: false,
+                storage: StorageQualifier::Input,
+            },
+            "gl_FrontFacing" => BuiltInData {
+                inner: TypeInner::Scalar {
                     kind: ScalarKind::Bool,
                     width: crate::BOOL_WIDTH,
                 },
-                BuiltIn::FrontFacing,
-                false,
-                StorageQualifier::Input,
-            ),
-            "gl_PrimitiveID" => add_builtin(
-                TypeInner::Scalar {
+                builtin: BuiltIn::FrontFacing,
+                mutable: false,
+                storage: StorageQualifier::Input,
+            },
+            "gl_PointSize" | "gl_FragDepth" => BuiltInData {
+                inner: TypeInner::Scalar {
+                    kind: ScalarKind::Float,
+                    width: 4,
+                },
+                builtin: match name {
+                    "gl_PointSize" => BuiltIn::PointSize,
+                    "gl_FragDepth" => BuiltIn::FragDepth,
+                    _ => unreachable!(),
+                },
+                mutable: true,
+                storage: StorageQualifier::Output,
+            },
+            "gl_ClipDistance" | "gl_CullDistance" => {
+                let base = self.module.types.fetch_or_append(
+                    Type {
+                        name: None,
+                        inner: TypeInner::Scalar {
+                            kind: ScalarKind::Float,
+                            width: 4,
+                        },
+                    },
+                    meta.as_span(),
+                );
+
+                BuiltInData {
+                    inner: TypeInner::Array {
+                        base,
+                        size: crate::ArraySize::Dynamic,
+                        stride: 4,
+                    },
+                    builtin: match name {
+                        "gl_ClipDistance" => BuiltIn::PointSize,
+                        "gl_CullDistance" => BuiltIn::FragDepth,
+                        _ => unreachable!(),
+                    },
+                    mutable: self.meta.stage == ShaderStage::Vertex,
+                    storage: StorageQualifier::Output,
+                }
+            }
+            "gl_BaseVertex"
```

Comment:
can we avoid listing them here by going `_` and then handling the `None` return internally?